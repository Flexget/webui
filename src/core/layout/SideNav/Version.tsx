import React, { useEffect, useState } from 'react';
import semver from 'semver-compare';
import { css } from '@emotion/core';
import { IconButton, Theme } from '@material-ui/core';
import HelpOutlineIcon from '@material-ui/icons/HelpOutline';
import Typography from '@material-ui/core/Typography';
import { useFlexgetAPI } from 'core/api';
import { useGlobalStatus } from 'core/status/hooks';

interface Props {
  className?: string;
}

interface VersionResponse {
  apiVersion: string;
  flexgetVersion: string;
  latestVersion: string;
}

const wrapper = (theme: Theme) => css`
  color: ${theme.palette.primary.main};
  opacity: 1;
`;

const useVersion = () => {
  const [version, setVersion] = useState<VersionResponse>();
  const [{ loading, error }, getVersion] = useFlexgetAPI<VersionResponse>('/server/version');
  useGlobalStatus(loading, error);

  useEffect(() => {
    const fn = async () => {
      const resp = await getVersion();
      if (resp.ok) {
        setVersion(resp.data);
      }
    };
    fn();
  }, [getVersion]);

  return { loading, version };
};

const Version: React.FC<Props> = ({ className }) => {
  const { loading, version } = useVersion();

  if (loading || !version) {
    // showProgress
    return null;
  }

  const { flexgetVersion, apiVersion, latestVersion } = version;

  return (
    <div css={wrapper} className={className}>
      <Typography>Version Info</Typography>
      <Typography>
        {`Flexget: ${flexgetVersion} `}
        {semver(latestVersion, flexgetVersion) !== 1 && (
          <IconButton href="https://flexget.com/ChangeLog" color="inherit" size="small">
            <HelpOutlineIcon />
          </IconButton>
        )}
      </Typography>
      <Typography>{`API: ${apiVersion}`}</Typography>
    </div>
  );
};

export default Version;
