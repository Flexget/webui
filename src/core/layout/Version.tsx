import React, { useEffect, useState } from 'react';
import semver from 'semver-compare';
import styled from '@emotion/styled';
import theme from 'theme';
import { IconButton } from '@material-ui/core';
import HelpOutlineIcon from '@material-ui/icons/HelpOutline';
import Typography from '@material-ui/core/Typography';
import { useFlexgetAPI } from 'core/api';
import { useGlobalStatus } from 'core/status/hooks';

interface Props {
  className?: string;
}

interface VersionResponse {
  apiVersion: string;
  flexgetVersion: string;
  latestVersion: string;
}

const Wrapper = styled.div`
  color: ${theme.palette.primary[800]};
`;

const Line = styled(Typography)`
  margin: 0;
`;

const useVersion = () => {
  const [version, setVersion] = useState<VersionResponse | undefined>();
  const [{ loading, error }, getVersion] = useFlexgetAPI<VersionResponse>('/server/version');
  useGlobalStatus(loading, error);

  useEffect(() => {
    const fn = async () => {
      const resp = await getVersion();
      if (resp.ok) {
        setVersion(resp.data);
      }
    };
    fn();
  }, [getVersion]);

  return { loading, version };
};

const Version: React.FC<Props> = ({ className }) => {
  const { loading, version } = useVersion();

  if (loading || !version) {
    // showProgress
    return null;
  }

  const { flexgetVersion, apiVersion, latestVersion } = version;

  return (
    <Wrapper className={className}>
      <Line>Version Info</Line>
      <Line>
        {`Flexget: ${flexgetVersion} `}
        {semver(latestVersion, flexgetVersion) !== 1 && (
          <IconButton href="https://flexget.com/ChangeLog" color="inherit" size="small">
            <HelpOutlineIcon />
          </IconButton>
        )}
      </Line>
      <Line>{`API: ${apiVersion}`}</Line>
    </Wrapper>
  );
};

export default Version;
