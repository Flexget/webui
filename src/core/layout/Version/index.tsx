import React from 'react';
import semver from 'semver-compare';
import styled from '@emotion/styled';
import { useFlexgetAPI } from 'utils/hooks/api';
import theme from 'theme';
import { IconButton } from '@material-ui/core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface Props {
  className?: string;
}

interface VersionResponse {
  apiVersion: string;
  flexgetVersion: string;
  latestVersion: string;
}

const Wrapper = styled.div`
  color: ${theme.palette.primary[800]};
`;

const Line = styled.p`
  margin: 0;
`;

const Version: React.FC<Props> = ({ className }) => {
  const { loading, error, data } = useFlexgetAPI<VersionResponse>('/server/version', []);

  if (loading || !data) {
    // showProgress
    return null;
  }

  if (error) {
    return null;
  }

  const { flexgetVersion, apiVersion, latestVersion } = data;

  return (
    <Wrapper className={className}>
      <Line>Version Info</Line>
      <Line>
        {`Flexget: ${flexgetVersion} `}
        {semver(latestVersion, flexgetVersion) === 1 && (
          <IconButton href="https://flexget.com/ChangeLog">
            <FontAwesomeIcon icon={['far', 'question-circle']} fixedWidth />
          </IconButton>
        )}
      </Line>
      <Line>{`API: ${apiVersion}`}</Line>
    </Wrapper>
  );
};

export default Version;
