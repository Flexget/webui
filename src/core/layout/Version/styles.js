import styled from '@emotion/styled';
import theme from 'theme';

export const Wrapper = styled.div`
  color: ${theme.palette.primary[800]};
`;

export const Line = styled.p`
  margin: 0;
`;
