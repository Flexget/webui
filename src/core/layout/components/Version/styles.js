import styled from 'react-emotion';
import theme from 'theme';

export const Wrapper = styled.div`
  color: ${theme.palette.error[500]};
`;

export const Line = styled.p`
  margin: 0;
`;
