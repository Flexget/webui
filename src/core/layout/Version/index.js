import { connect } from 'react-redux';
import actions from 'core/version/state/actions';
import Version from './Version';

export function mapStateToProps({ version }) {
  return {
    version,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    getVersion: () => dispatch(actions.getVersion.request()),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Version);
export { Version };
