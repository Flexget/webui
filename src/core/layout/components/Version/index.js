import { connect } from 'react-redux';
import { GET_VERSION } from 'core/version/data/actions';
import { request } from 'utils/actions';
import Version from './Version';

export function mapStateToProps({ version }) {
  return {
    version,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    getVersion: () => dispatch(request(GET_VERSION)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Version);
export { Version };
