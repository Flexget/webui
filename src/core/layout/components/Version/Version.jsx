import React, { Component } from 'react';
import PropTypes from 'prop-types';
import semver from 'semver-compare';
import IconButton from 'material-ui/IconButton';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import { Wrapper, Line } from './styles';

class Version extends Component {
  static propTypes = {
    version: PropTypes.shape({
      api: PropTypes.string,
      flexget: PropTypes.string,
      latest: PropTypes.string,
    }).isRequired,
    getVersion: PropTypes.func.isRequired,
    className: PropTypes.string,
  };

  static defaultProps = {
    className: '',
  };

  componentDidMount() {
    const { version, getVersion } = this.props;
    if (process.env.NODE_ENV === 'production' || !version.api) {
      getVersion();
    }
  }

  render() {
    const { version: { api, flexget, latest }, className } = this.props;
    return (
      <Wrapper className={className}>
        <Line>Version Info</Line>
        <Line>
Flexget:
          { flexget }
          {' '}
          {
          latest && semver(latest, flexget) === 1 && (
            <IconButton href="https://flexget.com/ChangeLog">
              <FontAwesomeIcon icon={['far', 'question-circle']} fixedWidth />
            </IconButton>
          ) }
        </Line>
        <Line>
API:
          { api }
        </Line>
      </Wrapper>
    );
  }
}

export default Version;
