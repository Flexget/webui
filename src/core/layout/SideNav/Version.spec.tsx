import React from 'react';
import fetchMock from 'fetch-mock';
import { cleanup } from '@testing-library/react';
import { renderWithWrapper } from 'utils/tests';
import Version from './Version';

describe('core/layout/Version', () => {
  afterEach(() => {
    cleanup();
    fetchMock.reset();
  });

  const component = <Version />;

  it('does not show icon with latest version', async () => {
    fetchMock.get('/api/server/version', {
      apiVersion: '1.1.2',
      flexgetVersion: '2.10.60',
      latestVersion: '2.10.60',
    });
    const { findByText, queryByLabelText } = renderWithWrapper(component);
    expect(await findByText('Flexget: 2.10.60')).toBeInTheDocument();
    expect(queryByLabelText('flexget update available')).not.toBeInTheDocument();
  });

  it('does not show icon with latest dev version', async () => {
    fetchMock.get('/api/server/version', {
      apiVersion: '1.1.2',
      flexgetVersion: '2.10.61.dev',
      latestVersion: '2.10.60',
    });
    const { findByText, queryByLabelText } = renderWithWrapper(component);
    expect(await findByText('Flexget: 2.10.61.dev')).toBeInTheDocument();
    expect(queryByLabelText('flexget update available')).not.toBeInTheDocument();
  });

  it('does show icon with a previous dev version', async () => {
    fetchMock.get('/api/server/version', {
      apiVersion: '1.1.2',
      flexgetVersion: '2.10.60.dev',
      latestVersion: '2.10.60',
    });
    const { findByText, queryByLabelText } = renderWithWrapper(component);
    expect(await findByText('Flexget: 2.10.60.dev')).toBeInTheDocument();
    expect(queryByLabelText('flexget update available')).toBeInTheDocument();
  });

  it('does show the icon with an old version', async () => {
    fetchMock.get('/api/server/version', {
      apiVersion: '1.1.2',
      flexgetVersion: '2.10.11',
      latestVersion: '2.10.60',
    });
    const { findByText, queryByLabelText } = renderWithWrapper(component);
    expect(await findByText('Flexget: 2.10.11')).toBeInTheDocument();
    expect(queryByLabelText('flexget update available')).toBeInTheDocument();
  });
});
