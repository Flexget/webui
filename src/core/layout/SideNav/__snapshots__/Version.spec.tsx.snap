// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/layout/Version renders correctly with latest version 1`] = `
<div
  css={[Function]}
>
  <p
    className="MuiTypography-root MuiTypography-body1"
  >
    Version Info
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1"
  >
    Flexget: 2.10.60 
    <a
      aria-disabled={false}
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall"
      href="https://flexget.com/ChangeLog"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </a>
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1"
  >
    API: 1.1.2
  </p>
</div>
`;

exports[`core/layout/Version renders correctly without latest version 1`] = `
<div
  css={[Function]}
>
  <p
    className="MuiTypography-root MuiTypography-body1"
  >
    Version Info
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1"
  >
    Flexget: 2.10.11 
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1"
  >
    API: 1.1.2
  </p>
</div>
`;
